### PR TITLE
feat: Frame Cryptor (aes gcm/cbc).

### PR DESCRIPTION
**Pull request by cloudwebrtc**
- [x] iOS/macOS
- [x] Android
- [x] Windows/Linux https://github.com/webrtc-sdk/libwebrtc/pull/46

### CODE REVIEW

File: api/crypto/frame_crypto_transformer.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+ public:
+  virtual const std::vector<std::vector<uint8_t>> keys(
+      const std::string participant_id) const = 0;
+
+ protected:
+  virtual ~KeyManager() {}
+};
+
+class FrameCryptorTransformer
```
Comment (theomonnom) [inline api/crypto/frame_crypto_transformer.h]: Maybe a small detail for now, can you export the class?
So I will be able to use it when compiling a static library

Comment (cloudwebrtc) [reply api/crypto/frame_crypto_transformer.h]: sure, I will do it.

Comment (cloudwebrtc) [reply api/crypto/frame_crypto_transformer.h]: https://github.com/webrtc-sdk/webrtc/pull/54/commits/9f1ebc27ee83dcb4a08026c5ea47e0be21377687

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+              std::vector<uint8_t> key) {
+    if (index > webrtc::KeyManager::kMaxKeySize) {
+      return false;
+    }
+
+    if (keys_.find(participant_id) == keys_.end()) {
+      keys_[participant_id] = std::vector<std::vector<uint8_t>>();
+    }
+
+    webrtc::MutexLock lock(&mutex_);
```
Comment (theomonnom) [inline sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: Shouldn't the lock be at the start of the function?

File: sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  bool SetKey(const std::string participant_id, int index, std::vector<uint8_t> key) {
+    if (index > webrtc::KeyManager::kMaxKeySize) {
+      return false;
+    }
+
+    if (keys_.find(participant_id) == keys_.end()) {
+      keys_[participant_id] = std::vector<std::vector<uint8_t>>();
+    }
+
+    webrtc::MutexLock lock(&mutex_);
```
Comment (theomonnom) [inline sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm]: Same here?

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+class DefaultKeyManagerImpl : public webrtc::KeyManager {
+ public:
+  DefaultKeyManagerImpl() = default;
+  ~DefaultKeyManagerImpl() override = default;
+
+  /// Set the key at the given index.
+  bool SetKey(const std::string participant_id,
+              int index,
+              std::vector<uint8_t> key) {
+    if (index > webrtc::KeyManager::kMaxKeySize) {
```
Comment (theomonnom) [inline sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: Is the the key size or the number of keys that should be limited to 32?

**Review by theomonnom — COMMENTED** [review]
This is 🔥
Just curious, why you didn't use [webrtc::FrameEncryptorInterface](https://github.com/webrtc-sdk/webrtc/blob/m104_release/api/crypto/frame_encryptor_interface.h) (I guess you have not enough flexibility with it?)

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h (hunk 2 above)
Comment (cloudwebrtc) [reply sdk/android/src/jni/pc/frame_cryptor_key_manager.h:45]: Yes, webrtc::FrameEncryptorInterface is missing the key information of FrameCrypto, codec type (VP8, H264 use different encryption methods), frame type, ssrc and timestamp (for building IV)

File: sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm (hunk 3 above)
Comment (cloudwebrtc) [reply sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm:39]: > This is 🔥 Just curious, why you didn't use [webrtc::FrameEncryptorInterface](https://github.com/webrtc-sdk/webrtc/blob/m104_release/api/crypto/frame_encryptor_interface.h) (I guess you have not enough flexibility with it?)

Yes, FrameEncryptorInterface is missing the key information of FrameCrypto, codec type (VP8, H264 use different encryption methods), frame type, ssrc and timestamp (for building IV)

**Comment by cloudwebrtc** [comment]
> This is 🔥 Just curious, why you didn't use [webrtc::FrameEncryptorInterface](https://github.com/webrtc-sdk/webrtc/blob/m104_release/api/crypto/frame_encryptor_interface.h) (I guess you have not enough flexibility with it?)

Yes, FrameEncryptorInterface is missing the key information of FrameCrypto, codec type (VP8, H264 use different encryption methods), frame type, ssrc and timestamp (for building IV)

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h (hunk 2 above)
Comment (cloudwebrtc) [reply sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: You are right, you should lock at the beginning of the function line

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h (hunk 4 above)
Comment (cloudwebrtc) [reply sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: the index should not be limited, but key bytes need to be limited to 32 (or just be 16 or 32)
https://github.com/webrtc-sdk/webrtc/pull/54/files/b6c80e4a573e216c4ad22652dd76c108d6780737#diff-7c5d6536fc3fee31501ca8f2f430869b0cac2a050abefe6b5077af-R6077af1d92ae,

image

So there should be some updates here

File: sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  bool SetKeys(const std::string participant_id, std::vector<std::vector<uint8_t>> keys) {
+    webrtc::MutexLock lock(&mutex_);
+    if (keys_.find(participant_id) == keys_.end()) {
+      keys_[participant_id] = std::vector<std::vector<uint8_t>>();
+    }
+
+    keys_[participant_id].clear();
+    for (auto key : keys) {
+      keys_[participant_id].push_back(key);
+    }
```
Comment (theomonnom) [inline sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm]:
```suggestion
    keys_[participant_id] = keys;
```

File: sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    if (keys_.find(participant_id) == keys_.end()) {
+      return std::vector<std::vector<uint8_t>>();
+    }
+
+    return keys_.find(participant_id)->second;
+  }
+
+ private:
+  mutable webrtc::Mutex mutex_;
+  std::map<std::string, std::vector<std::vector<uint8_t>>> keys_;
```
Comment (theomonnom) [inline sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm]:
```suggestion
  std::unordered_map<std::string, std::vector<std::vector<uint8_t>>> keys_;
```

Comment (theomonnom) [reply sdk/objc/api/peerconnection/RTCFrameCryptorKeyManager.mm]: Some really minor changes

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+               std::vector<std::vector<uint8_t>> keys) {
+    webrtc::MutexLock lock(&mutex_);
+    if (keys_.find(participant_id) == keys_.end()) {
+      keys_[participant_id] = std::vector<std::vector<uint8_t>>();
+    }
+
+    keys_[participant_id].clear();
+    for (auto key : keys) {
+      keys_[participant_id].push_back(key);
+    }
```
Comment (theomonnom) [inline sdk/android/src/jni/pc/frame_cryptor_key_manager.h]:
```suggestion
    keys_[participant_id] = keys;
```

**Review by theomonnom — APPROVED** [review]
lgtm!

File: sdk/android/src/jni/pc/frame_cryptor_key_manager.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    return true;
+  }
+
+  const std::vector<std::vector<uint8_t>> GetKeys(
```
Comment (theomonnom) [inline sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: Is there a reason why there are 2 getters here?

Comment (cloudwebrtc) [reply sdk/android/src/jni/pc/frame_cryptor_key_manager.h]: okay，removed the duplicate

**Pull request merged**